bundle/src/main/java/com/cognifide/cq/cqsm/core/utils/sling/SlingHelper.java

```diff
@@ -125,22 +135,53 @@ public static void operateTraced(ResourceResolverFactory factory, String userId,
 		}
 	}
 
-	/**
-	 * Create a new session for specified user (impersonating).
-	 */
-	public static ResourceResolver getResourceResolverForUser(ResourceResolverFactory factory, String userId)
-			throws LoginException {
-		ResourceResolver resolver;
-		if (userId != null) {
-			Map<String, Object> authenticationInfo = Maps.newHashMap();
-			authenticationInfo.put(ResourceResolverFactory.USER_IMPERSONATION, userId);
-			resolver = factory.getAdministrativeResourceResolver(authenticationInfo);
-		} else {
-			Map<String, Object> parameters = new HashMap<>();
-			parameters.put(ResourceResolverFactory.SUBSERVICE, SUBSERVICE_NAME);
-			resolver = factory.getServiceResourceResolver(parameters);
-		}
-
-		return resolver;
-	}
+    /**
+     * Create a new session for specified user (impersonating).
+     */
+    public static ResourceResolver getResourceResolverForUser(ResourceResolverFactory factory, String userId)
+            throws LoginException, RepositoryException {
+        Map<String, Object> parameters = new HashMap<>();
+        parameters.put(ResourceResolverFactory.SUBSERVICE, SUBSERVICE_NAME);
+        ResourceResolver resolver = factory.getServiceResourceResolver(parameters);
+        if (userId != null) {
+            Session session = resolver.adaptTo(Session.class);
+            addCqsmUserAsImpersonator(session, userId);
+            resolver = impersonate(factory, userId, session);
+        }
+
+        return resolver;
+    }
+
+    private static void addCqsmUserAsImpersonator(Session session, String userId) throws RepositoryException {
+        JackrabbitSession js = (JackrabbitSession) session;
+        UserManager userManager = js.getUserManager();
+        User user = ((User) userManager.getAuthorizable(userId));
+        if (!cqsmUserIsImpersonator(user)) {
+            User cqsmUser = ((User) userManager.getAuthorizable(CQSM_USER_ID));
+            user.getImpersonation().grantImpersonation(cqsmUser.getPrincipal());
+            session.save();
+        }
+    }
+
+    private static boolean cqsmUserIsImpersonator(User user) throws RepositoryException {
+        boolean isImpersonator = false;
+        PrincipalIterator impersonators = user.getImpersonation().getImpersonators();
+        while (impersonators.hasNext()) {
+            if (impersonators.nextPrincipal().getName().equals(CQSM_USER_ID)) {

```

Comment:
while comparing constants, always use `constant.equals(dynamicVar)` / by keeping this convention you will be always safe